### ISSUE TITLE
Bug on Discovery Devices

### ISSUE THREAD

**Issue by JohnnySun** [Freeyourgadget/Gadgetbridge#396]
On some devices (like xiaomi redmi note3 Medietek edition)
supportsService(UUID aService) returned `no cached services available for xxx`

heres some log 
`D/BluetoothLeScanner: onScanResult() - ScanResult{mDevice=D2:C8:48:09:71:CF, mScanRecord=ScanRecord [mAdvertiseFlags=6, mServiceUuids=[0000fee0-0000-1000-8000-00805f9b34fb, 0000fee7-0000-1000-8000-00805f9b34fb], mManufacturerSpecificData={343=[0, -68, -2, 1, 31, -70, 40, -20, 89, -109, -13, 8, -122, -104, 65, 27, -51, 0, -46, -56, 72, 9, 113, -49]}, mServiceData={0000fee0-0000-1000-8000-00805f9b34fb=[0, 0, 0, 0]}, mTxPowerLevel=-2147483648, mDeviceName=MI Band 2], mRssi=-94, mTimestampNanos=839780009464}`

`W/nodomain.freeyourgadget.gadgetbridge.impl.GBDeviceCandidate: no cached services available for MI Band 2: D2:C8:48:09:71:CF`

**Comment by JohnnySun** [comment]
At bluetoothLeScanner log, i see `mServiceUuids=[0000fee0-0000-1000-8000-00805f9b34fb, 0000fee7-0000-1000-8000-00805f9b34fb]`, but `ParcelUuid[] uuids = device.getUuids();` still return null.

**Comment by JohnnySun** [comment]
should we use `ParcelUuid[] uuids = device.getUuids();` to fliter devices? 
If use startBTLEDiscovery() instead of startBTDiscovery(), the system will only return ble devices.
and filter with ManufacturerData is a good idea.

**Comment by cpfeiffer** [comment]
Right,  I never got any service UUIDs during discovery either.

But it's a long story.

First, my experience is that the LE scan does not report any devices at all, or, only after scanning a few  times. I don't know why. However, when we perform the general BT scan, we do find both LE and non-LE devices.

Second, since we support devices from different manufacturers, we cannot filter for a single manufacturer. If we could filter for a set of manufacturers, we could do that. Alternatively, we can do the filtering on the client side (just like we try check for supported services), but since we do not get any of that advertising data, we have to use device names and mac addresses for filtering.

Maybe this will get better when we target a newer android version where we can use the newer LE scan API, but somehow I doubt it.

Alas, I didn't have the time to investigate why we do not get any of the advertising data. If you have any idea, please share :-)

**Comment by JohnnySun** [comment]
Hi
First, The LE Scan is slower than general BT Scan its normal, because LE Scan need to send scan quest for every le devices to get Scan response.

On my test, its all right to use BTLE Scan. adv data and response data receive correctly.

Second, there a new API for Le scan added on API21 [Click Here](https://developer.android.com/reference/android/bluetooth/le/BluetoothLeScanner.html#startScan(java.util.List<android.bluetooth.le.ScanFilter>, android.bluetooth.le.ScanSettings, android.bluetooth.le.ScanCallback)), maybe its more effective that older `startLeScan()` API.

**Comment by JohnnySun** [comment]
Maybe we can check the SDK_INT on running, if user is android L+, use new LeScan api instead of general BT Scan

**Comment by JohnnySun** [comment]
@cpfeiffer I had add new BTLE API to discovery code, on my test, oneplus2 whth cm13 and redmi note3 with miui8 can recevie advertising data and show device on the list view.

More info, see  #398

**Comment by JohnnySun** [comment]
#398

**Comment by cpfeiffer** [comment]
I needed to change the code to perform both discoveries again (atm we first do non-LE, then LE discovery), because e.g. the Pebble does non-LE bluetooth.

In the long term, maybe we can fully switch to LE, but for now, we have to do both. Maybe we need to change the order to first search for LE and then non-LE, because the non-LE discovery would also find LE devices, but without their services.

**Comment by timoc** [comment]
I have an [M2s,](https://www.tinydeal.com/m2s-smart-bracelet-call-sms-reminder-sports-tracker-sleep-monitor-p-163073.html) which is not picked up by the device scan, but can be paired by the phone. 

Can i suggest that you add (via the debugging interface?) a mechanism for connecting to an already paired device? (with option for product interface to emulate)
The benefit here is that users can explore the existing options and see what does and does not work with their device. Paired (pardon the pun) with some form of serial port logging, it could be handy in reverse engineering the protocol

**Comment by cpfeiffer** [comment]
Interesting device, looks a bit familiar ;-)

Glad you asked and interesting topic that deserves a separate ticket, actually.

1. There is an option to enter a Mac address of a Mi Band in the preferences. We don't really need that option there anymore, but it would indeed be helpful for any kind of new device, either in the Discovery or Debug Activity.

2. In my optimistic view (or naive, if you want) I envisioned at least partly supporting devices generically. The Bluetooth specifications are very detailed with many many standardized services. The services and characteristics of a device could be discovered dynamically and then directly used by Gadgetbridge.

I actually started implementing some of the basic Bluetooth services/profiles and intended to move them into a separate repository even, so that they can be reused independently of Gadgetbridge or even Android.

I don't have a single device that
- works without a custom initialization/authorization
- really makes use of standardized services
so the whole thing never really took off.

3. So far I've used https://github.com/ligi/BLExplorer for probing devices and their capabilities.

4. Interesting link -- need to find some time to really read up on it and think of a way to make use of it (do you already have an idea bout this?).

**Comment by timoc** [comment]
1. I used the mac address approach - and no luck. but it is a good option please keep :)

2. I am reading up on the GATT specs, and as such have no comment on how manageable a generic approach is (as yet). But there is some hope. After all, as my link shows, our whole technology base is predicated on copying and tweaking 'reference implementations' and then commoditizing them. They don't make these firmwares from scratch, there is likely 3 or 4 reference implementations of which there are 40-50 tweaked variants in the field. In the end, the technology will get too complicated for this kind of custom development of effort, so the services will standardise. Though i don't think we will ever have a situation where the init/auth will be open, as they consider the data their way of monetising us the user, and that we have free access to it means that we can compete against them for features.
  
3. just came across nrf connect, but will check this one out.

4. no it is in my RTFM list, but i don't need to understand how it works to use it to help reverse engineer the protocol vocabulary, especially when the protocol for a fitness band has such a small set of functions, and hence easier to de-construct - especially as form follows function (e.g. setting date/time will probably use the same data). If there is a C++ implementation it should be usable for this kind of work.

**Comment by cpfeiffer** [comment]
1. Yeah, but it should be possible to select the device to try (in your base, Mi Band 2, not 1). But a dedicated activity to inspect new devices, similar to BLExplorer would be nice. Maybe just extending BLExplorer might even be sufficient.

2. There's so many standards, but they only rarely used, at least for activity trackers and smart watches. See https://www.bluetooth.com/specifications/gatt and https://www.bluetooth.com/specifications/adopted-specifications

3. Please share your experience with it, ditto for 4.

**Comment by cpfeiffer** [comment]
Closing this, please open a separate ticket for the exploration options.
